airbyte-integrations/connectors/destination-bigquery/src/main/kotlin/io/airbyte/integrations/destination/bigquery/operation/BigQueryStorageOperation.kt

```diff
@@ -29,39 +30,91 @@ abstract class BigQueryStorageOperation<Data>(
     protected val datasetLocation: String
 ) : StorageOperation<Data> {
     private val existingSchemas = ConcurrentHashMap.newKeySet<String>()
-    override fun prepareStage(streamId: StreamId, destinationSyncMode: DestinationSyncMode) {
+    override fun prepareStage(streamId: StreamId, suffix: String, replace: Boolean) {
         // Prepare staging table. For overwrite, it does drop-create so we can skip explicit create.
-        if (destinationSyncMode == DestinationSyncMode.OVERWRITE) {
-            truncateStagingTable(streamId)
+        if (replace) {
+            truncateStagingTable(streamId, suffix)
         } else {
-            createStagingTable(streamId)
+            createStagingTable(streamId, suffix)
         }
     }
 
-    private fun createStagingTable(streamId: StreamId) {
-        val tableId = TableId.of(streamId.rawNamespace, streamId.rawName)
+    override fun overwriteStage(streamId: StreamId, suffix: String) {
+        if (suffix == "") {
+            throw IllegalArgumentException("Cannot overwrite raw table with empty suffix")
+        }
+        bigquery.delete(tableId(streamId, ""))
+        bigquery.query(
+            QueryJobConfiguration.of(
+                """ALTER TABLE `${streamId.rawNamespace}`.`${streamId.rawName}$suffix` RENAME TO `${streamId.rawName}`"""
+            ),
+        )
+    }
+
+    override fun transferFromTempStage(streamId: StreamId, suffix: String) {
+        if (suffix == "") {
+            throw IllegalArgumentException(
+                "Cannot transfer records from temp raw table with empty suffix"
+            )
+        }
+        // TODO figure out how to make this work
+        // something about incompatible partitioning spec (probably b/c we're copying from a temp
+        // table partitioned on generation ID into an old real raw table partitioned on
+        // extracted_at)
+        val tempRawTable = tableId(streamId, suffix)
+        //        val jobConf =
+        //            CopyJobConfiguration.newBuilder(tableId(streamId, ""), tempRawTable)
+        //                .setWriteDisposition(JobInfo.WriteDisposition.WRITE_APPEND)
+        //                .build()
+        //        val job = bigquery.create(JobInfo.of(jobConf))
+        //        BigQueryUtils.waitForJobFinish(job)
+
+        bigquery.query(
+            QueryJobConfiguration.of(
+                """
+                    INSERT INTO `${streamId.rawNamespace}`.`${streamId.rawName}`
+                    SELECT * FROM `${streamId.rawNamespace}`.`${streamId.rawName}$suffix`
+                """.trimIndent()
+            )
+        )
+        bigquery.delete(tempRawTable)
+    }
+
+    override fun getStageGeneration(streamId: StreamId, suffix: String): Long? {
+        val result: TableResult =
+            bigquery.query(
+                QueryJobConfiguration.of(
+                    "SELECT _airbyte_generation_id FROM ${streamId.rawNamespace}.${streamId.rawName}$suffix LIMIT 1"
```

Comment:
This doesn't yell in BigQuery if the table doesn't exist ?  throws exception in jdbc land with table not exists or unauthorized i think.